File: backend/app/schemas/log.py
"""
ログシステムのスキーマ定義

APIリクエスト/レスポンスのデータ構造を定義
"""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field

from app.models.log import (
    CompletedLogStatus,
    EmotionalValence,
    LogFragmentRarity,
)


# LogFragment スキーマ
class LogFragmentBase(BaseModel):
    """ログフラグメントの基本スキーマ"""

    action_description: str = Field(description="行動の詳細な記述")
    keywords: list[str] = Field(default_factory=list, description="キーワード")
    emotional_valence: EmotionalValence = Field(default=EmotionalValence.NEUTRAL, description="感情価")
    rarity: LogFragmentRarity = Field(default=LogFragmentRarity.COMMON, description="レアリティ")
    importance_score: float = Field(default=0.0, ge=0.0, le=1.0, description="重要度スコア")
    context_data: dict[str, Any] = Field(default_factory=dict, description="行動時の文脈情報")


class LogFragmentCreate(LogFragmentBase):
    """ログフラグメント作成スキーマ"""

    character_id: str
    session_id: str


class LogFragmentRead(LogFragmentBase):
    """ログフラグメント読み取りスキーマ"""

    id: str
    character_id: str
    session_id: str
    created_at: datetime

    model_config = {"from_attributes": True}


# CompletedLog スキーマ
class CompletedLogBase(BaseModel):
    """完成ログの基本スキーマ"""

    name: str = Field(description="ログの名前")
    title: Optional[str] = Field(default=None, description="称号")
    description: str = Field(description="ログの説明文")
    skills: list[str] = Field(default_factory=list, description="獲得したスキル")
    personality_traits: list[str] = Field(default_factory=list, description="性格特性")
    behavior_patterns: dict[str, Any] = Field(default_factory=dict, description="行動パターン")


class CompletedLogCreate(CompletedLogBase):
    """完成ログ作成スキーマ"""

    creator_id: str
    core_fragment_id: str
    sub_fragment_ids: list[str] = Field(default_factory=list, description="サブフラグメントのIDリスト")


class CompletedLogUpdate(BaseModel):
    """完成ログ更新スキーマ"""

    name: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    skills: Optional[list[str]] = None
    personality_traits: Optional[list[str]] = None
    behavior_patterns: Optional[dict[str, Any]] = None
    status: Optional[CompletedLogStatus] = None


class CompletedLogRead(CompletedLogBase):
    """完成ログ読み取りスキーマ"""

    id: str
    creator_id: str
    core_fragment_id: str
    contamination_level: float
    status: CompletedLogStatus
    created_at: datetime
    completed_at: Optional[datetime]

    model_config = {"from_attributes": True}

# 活動記録スキーマ
class LogActivityEntry(BaseModel):
    """ログの活動記録エントリー"""

    timestamp: datetime
    action: str
    details: dict[str, Any]
    impact_score: float = Field(default=0.0, ge=0.0, le=1.0, description="行動の影響度")

File: frontend/src/types/log.ts
// ログシステムの型定義（バックエンドのPydanticモデルと一致）

// 感情価
export type EmotionalValence = 'positive' | 'negative' | 'neutral'

// ログフラグメントのレアリティ
export type LogFragmentRarity =
  | 'common'
  | 'uncommon'
  | 'rare'
  | 'epic'
  | 'legendary'

// 完成ログのステータス
export type CompletedLogStatus =
  | 'draft'
  | 'completed'
  | 'contracted'
  | 'active'
  | 'expired'
  | 'recalled'

// ログフラグメント（ログの欠片）
export interface LogFragment {
  id: string
  characterId: string
  sessionId: string
  actionDescription: string
  keywords: string[]
  emotionalValence: EmotionalValence
  rarity: LogFragmentRarity
  importanceScore: number
  contextData: Record<string, unknown>
  createdAt: string
}

// 完成ログ
export interface CompletedLog {
  id: string
  creatorId: string
  coreFragmentId: string
  fragmentIds: string[]
  name: string
  title: string
  description: string
  skills: string[]
  personalityTraits: string[]
  behaviorPatterns: Record<string, unknown>
  contaminationLevel: number
  status: CompletedLogStatus
  createdAt: string
  completedAt?: string
}

// 完成ログの読み取り用型
export interface CompletedLogRead extends CompletedLog {
  created_at: string
}

// ログフラグメントの読み取り用型
export type LogFragmentRead = LogFragment

// API リクエスト/レスポンス用の型

// ログフラグメント作成リクエスト
export interface LogFragmentCreate {
  characterId: string
  sessionId: string
  actionDescription: string
  keywords: string[]
  emotionalValence: EmotionalValence
  contextData?: Record<string, unknown>
}

// 完成ログ作成リクエスト
export interface CompletedLogCreate {
  creatorId: string
  coreFragmentId: string
  subFragmentIds: string[]
  name: string
  title?: string
  description: string
  skills?: string[]
  personalityTraits?: string[]
  behaviorPatterns?: Record<string, unknown>
}

// ログ編纂時の計算結果
export interface LogCompilationPreview {
  contaminationLevel: number
  suggestedName: string
  suggestedTitle: string
  suggestedDescription: string
  extractedSkills: string[]
  extractedPersonalityTraits: string[]
  warningMessages?: string[]
}
